iphone/TitaniumKit/TitaniumKit/Sources/API/ObjcProxy.m

```diff
@@ -0,0 +1,235 @@
+  @try {
+    if (m_listeners == nil) {
+      m_listeners = [[NSMutableDictionary alloc] initWithCapacity:3];
+    }
+    JSManagedValue *managedRef = [JSManagedValue managedValueWithValue:callback];
+    [callback.context.virtualMachine addManagedReference:managedRef withOwner:self];
+    NSMutableArray *listenersForType = [m_listeners objectForKey:name];
+    if (listenersForType == nil) {
+      listenersForType = [[NSMutableArray alloc] init];
+    }
+    [listenersForType addObject:managedRef];
+    ourCallbackCount = [listenersForType count];
+    [m_listeners setObject:listenersForType forKey:name];
+  }
+  @finally {
+    pthread_rwlock_unlock(&m_listenerLock);
+    [self _listenerAdded:name count:ourCallbackCount];
+  }
+}
+
+- (void)removeEventListener:(NSString *)name withCallback:(JSValue *)callback
+{
+  pthread_rwlock_wrlock(&m_listenerLock);
+  NSUInteger ourCallbackCount = 0;
+  BOOL removed = false;
+  @try {
+    if (m_listeners == nil) {
+      return;
+    }
+    NSMutableArray *listenersForType = (NSMutableArray *)[m_listeners objectForKey:name];
+    if (listenersForType == nil) {
+      return;
+    }
+
+    NSUInteger count = [listenersForType count];
+    for (NSUInteger i = 0; i < count; i++) {
+      JSManagedValue *storedCallback = (JSManagedValue *)[listenersForType objectAtIndex:i];
+      JSValue *actualCallback = [storedCallback value];
+      if ([actualCallback isEqualToObject:callback]) {
+        // if the callback matches, remove the listener from our mapping and mark unmanaged
+        [actualCallback.context.virtualMachine removeManagedReference:storedCallback withOwner:self];
+        [listenersForType removeObjectAtIndex:i];
+        [m_listeners setObject:listenersForType forKey:name];
+        ourCallbackCount = count - 1;
+        removed = true;
+        break;
+      }
+    }
+  }
+  @finally {
+    pthread_rwlock_unlock(&m_listenerLock);
+    if (removed) {
+      [self _listenerRemoved:name count:ourCallbackCount];
+    }
+  }
+}
+
+- (BOOL)_hasListeners:(NSString *)type
+{
+  pthread_rwlock_rdlock(&m_listenerLock);
+  @try {
+    if (m_listeners == nil) {
+      return NO;
+    }
+    NSMutableArray *listenersForType = (NSMutableArray *)[m_listeners objectForKey:type];
+    if (listenersForType == nil) {
+      return NO;
+    }
+    return [listenersForType count] > 0;
+  }
+  @finally {
+    pthread_rwlock_unlock(&m_listenerLock);
+  }
+}
+
+- (void)_listenerAdded:(NSString *)type count:(int)count
+{
+  // for subclasses
+}
+
+- (void)_listenerRemoved:(NSString *)type count:(int)count
+{
+  // for subclasses
+}
+
+- (void)fireEvent:(NSString *)name withDict:(NSDictionary *)dict
+{
+  pthread_rwlock_rdlock(&m_listenerLock);
+  @try {
+    if (m_listeners == nil) {
+      return;
+    }
+    NSArray *listenersForType = [m_listeners objectForKey:name];
+    if (listenersForType == nil) {
+      return;
+    }
+    // FIXME: looks like we need to handle bubble logic/etc. See other fireEvent impl
```

Comment:
so for now I punted on this because none of the proxies I converted made use of the bubbling stuff. That's effectively a UI thing, as far as I can tell.

There's a lot of ugly C-API JSC code around TiBindingEvent.m to do the old event firing, with lots of various fireEvent methods in TiProxy too.

But from what I can tell, the bubbles/propagate stuff was really meant to bubble events up the view hierarchy by using `parentForBubbling` which I only see overrides in Ti.Ui.View and Ti.UI.TableViewRow.

So, I think for now we can continue to punt on this bubbling stuff - **but it's entirely possible I did miss some other code around setting default event values when not specified.**

Reply:
Yeah, that totally makes sense. Apart from UI there shouldn't be any need for event bubbling at all. I guess then it's fine to but this aside until we touch the UI stuff.